bpfd-operator/test/integration/suite_test.go

```diff
@@ -0,0 +1,193 @@
+	bpfdAgentImage    = os.Getenv("BPFD_AGENT_IMG")
+	bpfdOperatorImage    = os.Getenv("BPFD_OPERATOR_IMG")
+
+	existingCluster      = os.Getenv("USE_EXISTING_KIND_CLUSTER")
+	keepTestCluster      = func() bool { return os.Getenv("TEST_KEEP_CLUSTER") == "true" || existingCluster != "" }()
+	// TODO (astoycos) add this back after fixing bpfd-operator deletion issues
+	//keepKustomizeDeploys = func() bool { return os.Getenv("TEST_KEEP_KUSTOMIZE_DEPLOYS") == "true" }()
+
+	cleanup = []func(context.Context) error{}
+)
+
+const (
+	bpfdKustomize = "../../config/default"
+)
+
+func TestMain(m *testing.M) {
+	// check that we have the bpfd, bpfd-agent, and bpfd-operator images to use for the tests.
+	// generally the runner of the tests should have built these from the latest
+	// changes prior to the tests and fed them to the test suite.
+	if bpfdImage == "" || bpfdAgentImage == "" || bpfdOperatorImage == ""  {
+		exitOnErr(fmt.Errorf("BPFD_IMG, BPFD_AGENT_IMG, and BPFD_OPERATOR_IMG must be provided"))
+	}
+
+	ctx, cancel = context.WithCancel(context.Background())
+	defer cancel()
+
+	if existingCluster != "" {
+		fmt.Printf("INFO: existing kind cluster %s was provided\n", existingCluster)
+
+		// if an existing cluster was provided, build a test env out of that instead
+		cluster, err := kind.NewFromExisting(existingCluster)
+		exitOnErr(err)
+		env, err = environments.NewBuilder().WithExistingCluster(cluster).Build(ctx)
+		exitOnErr(err)
+	} else {
+		fmt.Println("INFO: creating a new kind cluster")
+
+		// to use the provided bpfd, bpfd-agent, and bpfd-operator images we will need to add
+		// them as images to load in the test cluster via an addon.
+		loadImages, err := loadimage.NewBuilder().WithImage(bpfdImage)
+		exitOnErr(err)
+		loadImages, err = loadImages.WithImage(bpfdAgentImage)
+		exitOnErr(err)
+		loadImages, err = loadImages.WithImage(bpfdOperatorImage)
+		exitOnErr(err)
+
+
+		// create the testing environment and cluster
+		env, err = environments.NewBuilder().WithAddons(certmanager.New(),loadImages.Build()).Build(ctx)
+		exitOnErr(err)
+
+		if !keepTestCluster {
+			addCleanup(func(context.Context) error {
+				cleanupLog("cleaning up test environment and cluster %s\n", env.Cluster().Name())
+				return env.Cleanup(ctx)
+			})
+		}
+
+		fmt.Printf("INFO: new kind cluster %s was created\n", env.Cluster().Name())
+
+		// deploy the BPFD Operator and revelevant CRDs
+		fmt.Println("INFO: deploying bpfd operator to test cluster")
+		exitOnErr(clusters.KustomizeDeployForCluster(ctx, env.Cluster(), bpfdKustomize))
+		// TODO (astoycos) add this back after fixing bpfd-operator deletion issues
+		// if !keepKustomizeDeploys {
+		// 	addCleanup(func(context.Context) error {
+		// 		cleanupLog("cleaning up bpfd operator")
+		// 		return clusters.KustomizeDeleteForCluster(ctx, env.Cluster(), bpfdKustomize)
+		// 	})
+		// }
+	}
+
+	exitOnErr(waitForBpfdReadiness(ctx, env))
+
+	exit := m.Run()
+
+	exitOnErr(runCleanup())
```

Comment:
Have you tested when an error is returned from `runCleanup()`? Because `exitOnErr()` will call `runCleanup()` again. Should the `exitOnErr()` be dropped here?

Reply:
This was intentional :) I.e if we fail to cleanup make sure we try again at least one more time and log any errors